src/Service/Services/OpenAPI/OpenApiDocumentor.cs

```diff
@@ -0,0 +1,708 @@
+        {
+            Dictionary<string, OpenApiSchema> schemas = new();
+
+            foreach (string entityName in _metadataProvider.EntityToDatabaseObject.Keys.ToList())
+            {
+                // Entities which disable their REST endpoint must not be included in
+                // the OpenAPI description document.
+                if (_runtimeConfig.Entities.TryGetValue(entityName, out Entity? entity) && entity is not null)
+                {
+                    if (entity.GetRestEnabledOrPathSettings() is bool restEnabled)
+                    {
+                        if (!restEnabled)
+                        {
+                            continue;
+                        }
+                    }
+                }
+
+                SourceDefinition sourceDefinition = _metadataProvider.GetSourceDefinition(entityName);
+                List<string> columns = /*sourceDefinition is null ? new List<string>() : */sourceDefinition.Columns.Keys.ToList();
+
+                // create component for FULL entity with PK.
+                schemas.Add(entityName, CreateComponentSchema(entityName, fields: columns));
+
+                // create component for entity with no PK
+                // get list of columns - primary key columns then optimize
+                foreach (string primaryKeyColumn in sourceDefinition.PrimaryKey)
+                {
+                    columns.Remove(primaryKeyColumn);
+                }
+
+                // create component for TABLE (view?) NOT STOREDPROC entity with no PK.
+                DatabaseObject dbo = _metadataProvider.EntityToDatabaseObject[entityName];
+                if (dbo.SourceType is not SourceType.StoredProcedure or SourceType.View)
+                {
+                    schemas.Add($"{entityName}_NoPK", CreateComponentSchema(entityName, fields: columns));
+                }
+            }
+
+            return schemas;
+        }
+
+        /// <summary>
+        /// Creates the schema object for an entity.
+        /// </summary>
+        /// <param name="entityName">Name of the entity.</param>
+        /// <param name="fields">List of mapped (alias) field names.</param>
+        /// <exception cref="DataApiBuilderException">Raised when an entity's database metadata can't be found.</exception>
+        /// <returns>Entity's OpenApiSchema representation.</returns>
+        private OpenApiSchema CreateComponentSchema(string entityName, List<string> fields)
+        {
+            if (!_metadataProvider.EntityToDatabaseObject.TryGetValue(entityName, out DatabaseObject? dbObject) || dbObject is null)
+            {
+                throw new DataApiBuilderException(
+                    message: "Entity's database object metadata not found.",
+                    statusCode: HttpStatusCode.InternalServerError,
+                    subStatusCode: DataApiBuilderException.SubStatusCodes.OpenApiDocumentAlreadyExists);
```

Comment:
Open Api document generation failure seems like a better choice
```suggestion
                    subStatusCode: DataApiBuilderException.SubStatusCodes.OpenApiDocumentGenerationFailure);
```

Reply:
updated